### ISSUE TITLE
Can't Find libexec/apache2/ folder

### ISSUE THREAD

**Issue by veget88** [Homebrew/homebrew-php#3601]
## Description

I was able to install php56 with brew install php56. Everything goes fine but no libexec folder is created, so I can't insert in httpd.conf the new php5_module. The same happens with php70.
Any advice?
## Context and Logs

| Parameter | Value |
| --- | --- |
| **OS X Version:** | 10.12 |
| **Homebrew Version:** | 1.0.2-49 |
| **Xcode Version:** | 8.0 |

**Comment by aurimasniekis** [comment]
> With the release of macOS Sierra the Apache module is now not built by default. If you want to build it on your system
you have to install php with the --with-apache option. See  brew options php56  for more details.

**Comment by veget88** [comment]
I've tried the --with-apache option but it returns 

> Sorry, I cannot run apxs. 

What should I do?

**Comment by aurimasniekis** [comment]
Maybe try with `--with-homebrew-apxs`

**Comment by veget88** [comment]
Nothing to do, but thanks for your reply.

**Comment by aurimasniekis** [comment]
http://stackoverflow.com/questions/38855651/php7-on-macos-sierra-beta-fails-to-install

**Comment by veget88** [comment]
Now I've a different error.
```
/usr/include/apache2/ap_hooks.h:39:10: fatal error: 'apr.h' file not found
#include "apr.h"
         ^
1 error generated.
```

I'm reading this https://github.com/GrahamDumpleton/mod_wsgi/issues/97 but the solution it's not working.

**Comment by MaciejJanyska** [comment]
I have exactly the same problem.

**Comment by javian** [comment]
Can you try now ? There were some fixes to the apache module in #3603 that was just merged.

**Comment by veget88** [comment]
Now it returns 
`Error: No available formula with the name "homebrew/dupes/zlib" (dependency of homebrew/php/php70)`

**Comment by javian** [comment]
@veget88 strange - the zlib dupe exists for me

**Comment by droid-Q** [comment]
same problem.
`if build.with?('httpd') || build.with?('httpd22')` maybe not works,always return `'/usr/sbin/apxs'`

**Comment by josegonzalez** [comment]
@veget88 can you verify that you have the `homebrew/dupes` tap locally?

**Comment by droid-Q** [comment]
I think this is wrong.
```
def apache_apxs
    if build.with?('httpd') || build.with?('httpd22')
      ['sbin', 'bin'].each do |dir|
        if File.exist?(location = "#{HOMEBREW_PREFIX}/#{dir}/apxs")
          return location
        end
      end
    else
      '/usr/sbin/apxs'
    end
  end
```
not should use sbin.

**Comment by josegonzalez** [comment]
@droid-Q what should it use if you do not want to build with httpd or httpd22?

**Comment by droid-Q** [comment]
@josegonzalez it is my fault.it is right.

**Comment by droid-Q** [comment]
@josegonzalez i met this issue ：
```
  ✝  /usr/local/bin  brew reinstall  php71 php71-mcrypt --with-apache
==> Reinstalling homebrew/php/php71 with --with-apache
==> Downloading https://downloads.php.net/~davey/php-7.1.0RC2.tar.bz2
######################################################################## 100.0%
==> ./configure --prefix=/usr/local/Cellar/php71/7.1.0-rc.2_6 --localstatedir=/usr/local/var --sysconfdir=/usr/local/etc/php/7.1 --with-config-file-path=/usr/local/etc/php/7.1 --with-config-file-scan-di
Last 15 lines from /Users/droid/Library/Logs/Homebrew/php71/01.configure:
checking for pthreads_lib...

Configuring SAPI modules
checking for Apache 2.0 handler-module support via DSO through APXS...

Sorry, I cannot run apxs.  Possible reasons follow:

1. Perl is not installed
2. apxs was not found. Try to pass the path using --with-apxs2=/path/to/apxs
3. Apache was not built using --enable-so (the apxs usage page is displayed)

The output of /usr/sbin/apxs follows:

apxs:Error: /Applications/Xcode.app/Contents/Developer/Toolchains/OSX10.12.xctoolchain/usr/local/bin/apu-1-config not found!.
configure: error: Aborting

READ THIS: https://git.io/brew-troubleshooting
If reporting this issue please do so at (not Homebrew/brew):
  https://github.com/Homebrew/homebrew-php/issues
```

**Comment by p3palazzo** [comment]
I got the same problem with apache not able to find php 7's libexec folder, but on 10.11, so the build of Apache is not at issue here. The problem started when a recent Homebrew update moved the install location, but I can't put my finger on where libphp7.so went.

**Comment by javian** [comment]
@droid-Q : that's a different issue which should be fixed after #3626 is merged

**Comment by javian** [comment]
The fix has now been merged, do a `brew update` and then try to reinstall with `--with-apache` and it should work now.

**Comment by gigmaps** [comment]
got the same issue after running `brew update` then `brew upgrade` this morning... on a working version of php70 - i think

**Comment by javian** [comment]
@gigmaps you have to install php70 from source to fix it but you need to do that to compile the apache module anyhow since its no longer built by default. `brew install php70 --with-apache`

**Comment by gigmaps** [comment]
@javian thanks - note that i'm on 11.11.6 and had a working version of php70 prior to `brew upgrade`

also, in `/usr/local/Cellar/php70/` I have 2 versions of 7.0.11
- `/usr/local/Cellar/php70/7.0.11_2` created 2016-09-13 which includes the `libexec` folder
- `/usr/local/Cellar/php70/7.0.11_3` created 2016-09-26 which does not include the `libexec` folder

running `brew reinstall php70 -- with-apache --build-from-source` doesn't complete & gives the following error:
`Warning!  dlname not found in /usr/local/Cellar/php70/7.0.11_3/libexec/apache2/libphp7.so.
Assuming installing a .so rather than a libtool archive.
chmod 755 /usr/local/Cellar/php70/7.0.11_3/libexec/apache2/libphp7.so
apxs:Error: Config file /usr/local/etc/apache2/2.4/httpd.conf not found.
make: *** [install-sapi] Error 1`

**Comment by chrisdlangton** [comment]
same for me here

tried `brew upgrade php70 --with-apache` which after quite a long time succeeded in upgrading to `7.0.11` but when i inspect my directory `ls /usr/local/Cellar/php70/7.0.11_3/` all i get is;

```
INSTALL_RECEIPT.json       bin/                       php/
LICENSE                    homebrew.mxcl.php70.plist  sbin/
NEWS                       include/                   share/
README.md                  lib/                       var/
```

No `libexec` still

**Comment by grigutis** [comment]
I was looking for the libexec folder too. I'm on OS X 10.11.6 and tried the `build-from-source` option mentioned previously, but it doesn't seem to help. `brew install php70 --with-apache --build-from-source` produces the following:

```
Installing PHP SAPI module:       apache2handler
/usr/local/Cellar/httpd24/2.4.23_1/build/instdso.sh SH_LIBTOOL='/usr/local/opt/apr/libexec/build-1/libtool' libs/libphp7.so /usr/local/Cellar/php70/7.0.11_3/libexec/apache2
/usr/local/opt/apr/libexec/build-1/libtool --mode=install install libs/libphp7.so /usr/local/Cellar/php70/7.0.11_3/libexec/apache2/
libtool: install: install libs/libphp7.so /usr/local/Cellar/php70/7.0.11_3/libexec/apache2/libphp7.so
Warning!  dlname not found in /usr/local/Cellar/php70/7.0.11_3/libexec/apache2/libphp7.so.
Assuming installing a .so rather than a lib tool archive
chmod 755 /usr/local/Cellar/php70/7.0.11_3/libexec/apache2/libphp7.so
apxs:Error: Config file /usr/local/etc/apache2/2.4/httpd.conf not found.
make: *** [install-sapi] Error 1.
```

**Comment by chrisdlangton** [comment]
after force uninstall of all php I retried using `brew update` which was good, then `brew upgrade` which installed node and some other bits, then `brew install php70 --with-apache --build-from-source` which failed with the following output;

```
==> Installing php70 from homebrew/php
==> Downloading https://php.net/get/php-7.0.11.tar.bz2/from/this/mirror
Already downloaded: /Users/chrislangton/Library/Caches/Homebrew/php70-7.0.11
Warning: Backing up all known pear.conf and .pearrc files
Warning: If you have a pre-existing pear install outside
         of homebrew-php, or you are using a non-standard
         pear.conf location, installation may fail.
==> ./configure --prefix=/usr/local/Cellar/php70/7.0.11_3 --localstatedir=/usr/local/var --sysconfd
==> make
==> make install
Last 15 lines from /Users/chrislangton/Library/Logs/Homebrew/php70/03.make:
2016-09-27 11:31:38 +1000

make install

Installing PHP SAPI module:       apache2handler
/usr/local/Cellar/httpd24/2.4.23_1/build/instdso.sh SH_LIBTOOL='/usr/local/opt/apr/libexec/build-1/libtool' libs/libphp7.so /usr/local/Cellar/php70/7.0.11_3/libexec/apache2
/usr/local/opt/apr/libexec/build-1/libtool --mode=install install libs/libphp7.so /usr/local/Cellar/php70/7.0.11_3/libexec/apache2/
libtool: install: install libs/libphp7.so /usr/local/Cellar/php70/7.0.11_3/libexec/apache2/libphp7.so
Warning!  dlname not found in /usr/local/Cellar/php70/7.0.11_3/libexec/apache2/libphp7.so.
Assuming installing a .so rather than a libtool archive.
chmod 755 /usr/local/Cellar/php70/7.0.11_3/libexec/apache2/libphp7.so
apxs:Error: Config file /usr/local/etc/apache2/2.4/httpd.conf not found.
make: *** [install-sapi] Error 1

READ THIS: https://git.io/brew-troubleshooting
If reporting this issue please do so at (not Homebrew/brew):
  https://github.com/Homebrew/homebrew-php/issues

These open issues may also help:
Enabling php70-opcache https://github.com/Homebrew/homebrew-php/issues/3619
php70: update bottle https://github.com/Homebrew/homebrew-php/pull/3627
Problem installing php70-mcrypt, php70-opcache, php70-xdebug on El Capitan https://github.com/Homebrew/homebrew-php/issues/3587
Add php70-zmq formula https://github.com/Homebrew/homebrew-php/pull/3474
PHP70 does not create php.ini file https://github.com/Homebrew/homebrew-php/issues/3599
Problem installing homebrew/php/php70-imagick https://github.com/Homebrew/homebrew-php/issues/3571
Allow --enable-redis-igbinary for php70-redis https://github.com/Homebrew/homebrew-php/pull/3473
Add head formula for php70-uploadprogress https://github.com/Homebrew/homebrew-php/pull/3178
```

My `brew gist-logs php70`;
https://gist.github.com/182a164e4670d4be1d5b4200651fe3de

And I'm on 10.11.6 (15G1004)

**Comment by grigutis** [comment]
[It looks like `httpd24` is the real cause of the problem](https://github.com/Homebrew/homebrew-apache/issues/136). Installing httpd24 via:

    $ brew install -v https://raw.githubusercontent.com/Homebrew/homebrew-apache/7f75f548c89ee0bffab92e75399ee00ddf727a0c/httpd24.rb

And then

    $ brew install php70 --with-apache

produces a working `/usr/local/opt/php70/libexec/apache2/libphp7.so`

**Comment by Saeven** [comment]
Thanks I'll give this a shot!

**Comment by chrisdlangton** [comment]
@grigutis that manual process seems to have apache and php talking, now i'm just struggline with app environment configurations :/ ty

**Comment by Saeven** [comment]
Indeed, when httpd24 installs it overwrites the default (now heavily configured) configuration.

I don't know what led to this decision, to exclude libphp*.so from the "standard" brew install.  It'd sure be nice if this behavior were reverted, or an option given to "just install" like it used to. --with-so without installing the entire httpd24 config and local binary disaster.

**Comment by grigutis** [comment]
I agree. I'm using OS X Server's httpd, but Homebrew's php70. I really don't need or want Homebrew's httpd if I can help it.

**Comment by javian** [comment]
The httpd24 bottle has been updated which should create the correct directory structure.

@grigutis currently to my knowledge it is not possible to compile the apache module on macOS Sierra unless you make manual updates to your own system which has been done by others in other issues. This method can not be enforced through homebrew. The reason for this is that Apple removed the apr-1-config script from the OS.

There are a few options that can possibly be explored like providing a patched apxs script through the formula, create a seperate formula just for the php module which can be bottled or add support for doing it for systems older than Sierra. Feel free to make a contribution to make it work.

**Comment by JuanLuisGarciaBorrego** [comment]
Is there any solutions?
Thanks!

**Comment by leewillis77** [comment]
I hit this problem with PHP 7.0 on OSX Yosemite. I was able to get around it by installing PHP 7 with:

```
$ brew install --with-apache php70
$ brew unlink httpd24
```

No obvious problems spotted yet - YMMV...

**Comment by eppfel** [comment]
Tried the same, but then I hit #3633 .

**Comment by Kyslik** [comment]
@leewillis77 thanks will share you in #3633.

**Comment by black-snow** [comment]
Running OS X 10.11.6, had working versions of php54, 56 and 70, did `brew upgrade` and now nothing's working anymore. How can you just blow up existing setups?

Will try the apache solution now.

**Comment by byCedric** [comment]
I get it that the apache files are not generated anymore when installing PHP and you have to specifically declare you want the files. But why does it regenerate the files when upgrading? Just like @black-snow said even if you declared you want it, the `brew upgrade` breaks everything...

**Comment by Kyslik** [comment]
What OSX?

Install with apache and untap apache afterwards see 1-2 comments above.

**Comment by JuanLuisGarciaBorrego** [comment]
Is there any news?

**Comment by byCedric** [comment]
@Kyslik I upgraded to MacOS sierra, but it damaged a lot of other programs too. I did a reinstall and awaiting for a new PHP version to see if it still "removes" the apache support. Thanks for your quick response!

**Comment by Kyslik** [comment]
Bad move mate, I still use 10.10 because I am afraid of drastic changes Apple introduced...

And no problem. As I understand on sierra you have to install Apache with php because apple does not ship apache with sierra.

**Comment by impdesigns** [comment]
This problem was resolved for me once I realized I'd not installed the command line tools from upgraded Xcode.

**Comment by p3palazzo** [comment]
Reinstalling `php70 --with-apache` followed by `brew unlink httpd24` worked for me. 

--
Pedro Paulo Palazzo
Professor adjunto (Assistant professor)
Laboratório de Estudos da Urbe (LabeUrbe)
Universidade de Brasília
http://brasilia.academia.edu/p3palazzo

**Comment by byCedric** [comment]
After the update for homebrew's PHP `7.0.12_5` my `libphp7.so` disappeared once again. Even though I removed both apache and php from my system and reinstalled it with `php70 --with-apache`. As a side note, I was using the homebrew provided apache (not OSX's apache). I'm also now using MacOS's Sierra. Here is a [gist](https://gist.github.com/byCedric/4474ffcdc4c56449f6408c0a1cead80b) I created from my terminal. 

Probably reinstalling Apache with PHP once again will solve it. But having to do this over and over again for each update really annoys me. Is this expected behaviour?

**Comment by sarman** [comment]
--with-httpd24

**Comment by msnatu** [comment]
@sarman simple change... worked great! Thanks!

**Comment by cordominic** [comment]
Warning bro,  --with-apache was deprecated; using --with-httpd24 instead!

**Comment by markl181** [comment]
I attempted to upgrade from php56 to php70, and now having the same problem. I tried @grigutis suggestion, but that didn't create the .so file. Also tried brew install php70 --with-apache--with-pear--with-httpd24 and that didn't work either. Has anyone tried downgrading? I am tempted to give that a try since I really need a working php install.

**Comment by javian** [comment]
@markl181 can you open up a new issue for your problem and post a `brew gists-logs php70` there ?

**Comment by aurimasniekis** [comment]
Don't know. I am not using Apache.

**Comment by black-snow** [comment]
@Tom29 just install php `--with-httpd24` (install httpd24 before, oc) and then unlink httpd24 and use system apache

**Comment by Tom29** [comment]
@black-snow how can i unlink httpd24 and use system apache?
after using `brew install php56 --with-apache` how i can't stop apache by `apachectl stop`

**Comment by black-snow** [comment]
@Tom29 Use `brew services` ([link](https://github.com/Homebrew/homebrew-services)) to manage brew-installed services. Should be `$ brew services stop httdp24`.

You can unlink httpd24 via `$ brew unlink httpd24`

**Comment by spalburg** [comment]
due to apple macos sierra update my install broke
I've reinstalled my php71 --with-httpd24 and checked my apache httpd.conf LoadModulePHP that it loaded the right libphp7.so situated in usr/local/Cellar/php71/7.1.1_12/libexec/apache2 after this everything worked fine

**Comment by rosiecakes** [comment]
After a day of troubleshooting, @grigutis's suggestion worked to get my libexec back

**Comment by geeksweep** [comment]
I cannot get this to work... downloaded latest php 7.1.2.13 from homebrew. I have tried multiple ways (all comments/suggestions here) and nothing works. it looks like it also creates a different/added subfolder in the Cellar directory:

cd /usr/local/Cellar/php71/7.1.2_13/

and i get 

```
.brew/                     bin/                       sbin/
INSTALL_RECEIPT.json       homebrew.mxcl.php71.plist  share/
LICENSE                    include/                   var/
NEWS                       lib/                       
README.md                  php/ 

```

Im running El Capitan, OSX 10.11.6

any pointers/questions? Ill be glad to look through things to help out.

**Comment by black-snow** [comment]
@geeksweep `cd` prints your directory contents? :P
What *exactly* is your issue?

**Comment by geeksweep** [comment]
@black-snow haha yeah i was cd'ing then 'tabbing' to get those results -- sorry for the confusion. 

I cannot get the libexec folder to be created nor the shared object(as also mentioned in the comment before). I am attempting to set up a php web service for an iOS app - so that is why i am here.

**Comment by black-snow** [comment]
@geeksweep On a mac? Guess it's for development? In case you don't need fancy stuff you could also use the [built-in webserver](http://php.net/manual/en/features.commandline.webserver.php) (php5.4+). `$ cd` into your *doc root* and `$ php -S localhost:8080` or whichever* port you want.

Otherwise let's make a clean attempt.

```bash
$ brew uninstall --force php71
$ brew uninstall --force httpd24
$ brew cleanup
$ brew doctor # fix anything that pops up
$ brew install httpd24
$ which httpd24 # should be the Cellar one, if not, brew link httpd24 / troubleshoot
$ brew install --with-httpd24 php71 # or --with-apache, both may work
$ php -v # should be 7.1.x
$ brew unlink httpd24
```

**Comment by geeksweep** [comment]
@black-snow , Yes Mac OSX 10.11.6, El Capitan. This server will eventually live on a remote Mac and the iOS app will communicate with it to get/update/whatever on a mysql database.  (i assume phpmyadmin. thats a little bit out of topic but thats the goal. I dont have any back end service experience at all but im trying to learn the basics just to get by. The php web service will probably have about 5 or 6 unique calls (a mix of get and post) and thats about it. 
  
at any rate first warning i get here:

<img width="509" alt="screen shot 2017-02-22 at 8 38 34 pm" src="https://cloud.githubusercontent.com/assets/1126840/23241050/f6aa27b2-f93e-11e6-8a3a-b0cdec714a0e.png">

want to take this step by step in case i miss something, and then go from there. Thanks so much for your help in advance.

**Comment by p3palazzo** [comment]
Given that Apple has a knack for breaking the web server configuration with every system upgrade, it makes sense to use homebrew php and Apache.

**Comment by black-snow** [comment]
@p3palazzo true

@geeksweep yep, you can ignore the warning or just delete the header file (unless you put it there yourself for any reason).

**Comment by yunwe** [comment]
I followed all steps mentioned here, and I got working version of php7.1 together with libphp7.so. 
But when I add `LoadModule php7_module /usr/local/opt/php71/libexec/apache2/libphp7.so` to httpd.conf file, php files stop working. When I uncommented back `LoadModule php5_module libexec/apache2/libphp5.so` everything works fine. Any Idea?
PS. I installed php7.1 using `brew install php71 --with-httpd24` and unlinked httpd24
      OS. 10.12.2

**Comment by spalburg** [comment]
go over the brew upgrade php7.1 messages in your terminal it might seem that it upgraded but you might have some instructions that you missed to make it work

**Comment by spalburg** [comment]
furthermore before upgrading your php7.1 you probably had to unlink the php that was linked before upgrading and after the upgrade you have to link the proper php version for it to work

regards M

**Comment by spalburg** [comment]
brew doctor should tell you a thing or two about your php
did you actually check the paths you mentioned in your post with finder to see if the libphp files actually reside in that directory?
if you restart your apache server with: sudo apachectl -k restart do you get a message ?

**Comment by spalburg** [comment]
if you still cant get it to work after checking all that read the link i send you closely that you understand all that steps and compare it to your setup and adjust where needed.

https://getgrav.org/blog/macos-sierra-apache-multiple-php-versions <https://getgrav.org/blog/macos-sierra-apache-multiple-php-versions>

regards M

**Comment by spalburg** [comment]
last thing i can think of i NEVER had to unlinked my apache24 OS 10.12.2 for it to work
and my httpd.conf file resides in: /usr/local/etc/apache2/2.4

**Comment by yunwe** [comment]
I've been using system apache and got stuck there. After I've been following every steps in [https://getgrav.org/blog/macos-sierra-apache-multiple-php-versions](https://getgrav.org/blog/macos-sierra-apache-multiple-php-versions) and shifted to brew apache, it works like a charm. Thank you.

**Comment by aaronbauman** [comment]
For anyone still landing here:
```
--with-homebrew-apxs was deprecated; use --with-httpd instead!
--with-apache was deprecated; use --with-httpd instead!
--with-httpd24 was deprecated; use --with-httpd instead!
```

**Comment by viktym** [comment]
Hi @aaronadamsTO 
Using option --with-httpd I always get php70 installed in TS mode ( Build with thread safety)
I really need php70 installed in NTS
How can I do this?

**Comment by SMillerDev** [comment]
There's a bug currently where installing with apache forces the build into threadsafe mode.

**Comment by viktym** [comment]
@SMillerDev is an issue created for this bug? Do you have some estimates when it can be fixed?

**Comment by josegonzalez** [comment]
@viktym please file a separate  issue rather than commenting on a months old bug that has nothing to do with your issue. You can compile your php70 without thread safety by building from source iirc.
